src/persist/src/indexed/columnar.rs

```diff
@@ -220,42 +346,43 @@ impl ColumnarRecordsBuilder {
     /// The guesses for key and val sizes are best effort, and if they end up being
     /// too small, the underlying buffers will be resized.
     pub fn reserve(&mut self, additional: usize, key_size_guess: usize, val_size_guess: usize) {
-        self.records.key_offsets.reserve(additional);
-        self.records.key_data.reserve(additional * key_size_guess);
-        self.records.val_offsets.reserve(additional);
-        self.records.val_data.reserve(additional * val_size_guess);
-        self.records.timestamps.reserve(additional);
-        self.records.diffs.reserve(additional);
+        self.key_offsets.reserve(additional);
+        self.key_data.reserve(additional * key_size_guess);
+        self.val_offsets.reserve(additional);
+        self.val_data.reserve(additional * val_size_guess);
+        self.timestamps.reserve(additional);
+        self.diffs.reserve(additional);
+        debug_assert_eq!(self.borrow().validate(), Ok(()));
     }
 
     /// Add a record to Self.
     pub fn push(&mut self, record: ((&[u8], &[u8]), u64, isize)) {
         let ((key, val), ts, diff) = record;
-        self.records.key_offsets.push(self.records.key_data.len());
-        self.records.key_data.extend_from_slice(key);
-        self.records.val_offsets.push(self.records.val_data.len());
-        self.records.val_data.extend_from_slice(val);
-        self.records.timestamps.push(ts);
-        self.records.diffs.push(diff);
-        self.records.len += 1;
-        // The checks on `key_offsets` and `val_offsets` are different than in the
-        // rest of the file because those arrays have not received their final
-        // value yet in finish().
-        debug_assert_eq!(self.records.key_offsets.len(), self.records.len);
-        debug_assert_eq!(self.records.val_offsets.len(), self.records.len);
-        debug_assert_eq!(self.records.timestamps.len(), self.records.len);
-        debug_assert_eq!(self.records.diffs.len(), self.records.len);
+        self.key_data.extend_from_slice(key);
+        self.key_offsets
+            .push(i32::try_from(self.key_data.len()).expect("batch is smaller than 2GB"));
```

Comment:
hmmm I'm a little bit hesitant about the 2 GB size limit, mostly because the rest of the code assumes batches can be as large as they need to. 

Concretely, sidestepping that I have 2 questions:

1. can we make this a u32 and by so doing bump the limit to 4 GB?
2. nit: the expects say "batch is smaller than 2GB" but really its asserting that the length of all keys is less than 2 GB (and same for all values below). Not sure how to reword the expect / if thats worth doing.

Reply:
> hmmm I'm a little bit hesitant about the 2 GB size limit, mostly because the rest of the code assumes batches can be as large as they need to.

we can avoid it by using i64, but I'd rather do the work of limiting the size of our batches, which we've long said we want to do anyway

> can we make this a u32 and by so doing bump the limit to 4 GB?

sadly, nope. parquet avoids unsigned types because they're hard to deal with on the jvm

> nit: the expects say "batch is smaller than 2GB" but really its asserting that the length of all keys is less than 2 GB (and same for all values below). Not sure how to reword the expect / if thats worth doing.

i had the same thought and the same conclusion (dunno what else to do). i went with the batch phrasing with an eye toward making 2GB our batch size limit, but it's also easy enough to make it more correct now and change the message later once we actually do the work